### ISSUE TITLE
[Question] Retrieval Model NaN Training Loss

### ISSUE THREAD

**Issue by kmcmearty** [tensorflow/recommenders#502]
I am having an issue with my retrieval modeling resulting in NaN loss. I have 2 datasets, one small (115k interactions) and one larger (1.36 M interactions). I developed my model / code around the small dataset and it is able to train, albeit the factorized topK is very low. 

When I run the model on my large dataset, the loss will NaN out after the first or second training step in the first epoch. I've tried some tricks to make the model learn like reducing the batch size or learning rate but this only results in holding off the NaN till the 2nd or 3rd step. 

I'm not sure if the issue is with my dataset or training features? I used a data drift detection tool from Evidently.AI and it used the Jensen-Shannon statistics to determine that Cus_Id, Item_ID, Item_Desc and Brand_Name had drifted between my small and large dataset. This makes me wonder if it is an issue with my dataset since the small one will train and the larger dataset will not.

On the other hand, I wonder if having only categorical features is causing the model to not train?

Could anyone take a look at my setup below and offer some suggestions on how I could make my model learn on a larger dataset?
<br>

My current retrieval architecture is as follows:

| Query_Embeddings_Tower | Candidate_Embeddings_Tower |
| :---| :---|
| User_ID (32 dims) | Item_ID (32 dims)
| Time_Stamp - normalized (32 dims) | Brand_Name - Vectorzied (32 dims)
| | Item_Description - Vectorized (32 dims)|
| | Item_Dept (MGP Name) - Vectorized (32 dims) |
| | Item_Sub_Dept (CGP Name) - Vectorized (32 dims)|

<u>Hyperparameters<u>

Query Tower Layer: 16 Dense
Candidate Tower Layer: 16 Dense
Loss Function / Metrics: Factorized Top K / Top categorical accuracy
Retrieval Optimizer: Adagrad
Learning Rate: 0.001
Training / Test Batch Size: 512
Item Catalog Count: 25,849 Candidates

<br>

Large Dataset Description
![image](https://user-images.githubusercontent.com/10493854/173375958-4ae644fb-ed82-41e7-afa1-649a2c198e14.png)
Rows: 1.36 M
Shoppers: 60.9k unique
Items: 21,774
Minimum Customer Purchase Count: 10

**Comment by patrickorlando** [comment]
Hey @kmcmearty,
It's mostly likely to do with the dataset and or your categorical lookups.

If you are using candidate sampling probability for bias correction and you have some candidates with `p < 1e-6` then I would clip the probability. This may already be addressed in a more recent release of the library.

Otherwise pass a few batches of your training data through the model manually and look for which examples end up with NaN values. It could be from incomplete vocabularies if you have not set any out of vocabulary tokens.

**Comment by siyu1992** [comment]
I've met the same issue, and I am also comfused whether the nan loss comes from the prediction score. I found this issue while I am using multiple item features.

I don't know how to fix this. @patrickorlando Could you please take a look at this, thanks !

My current retrieval architecture is as follows:

<html>
<body>


Query_Embeddings_Tower | Candidate_Embeddings_Tower
-- | --
user_embedding (64dims) | item_embedding (64 dims)
provice_code_embedding (4dims) | item_points - Normalized (1 dim) 
register_channel - Normalized (1 dim) | item_order_count- Normalized (1 dim)
user_brand_code - Normalized (1 dim) | item_view_count - Normalized (1 dim)
user_star_level_code - Normalized (1 dim) | item_order_amount - Normalized (1 dim)
  | item_view_amount - Normalized (1 dim)

</body>
</html>


The training log pieces are as follows:

`Epoch 396/400
538/538 - 79s - factorized_top_k/top_1_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_5_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_10_categorical_accuracy: 0.0000e+00 
- factorized_top_k/top_50_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_100_categorical_accuracy: 0.0000e+00 - loss: nan - regularization_loss: 0.0000e+00 - total_loss: nan
2022-06-14 05:21:32.223640: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 681613 of 1300000
2022-06-14 05:21:42.223630: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 1024823 of 1300000
2022-06-14 05:21:46.974571: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:228] Shuffle buffer filled.
2022-06-14 05:21:46.974672: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 1 of 1100000
2022-06-14 05:21:46.974707: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 2 of 1100000
2022-06-14 05:21:52.209707: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 542338 of 1100000
2022-06-14 05:21:58.152252: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:228] Shuffle buffer filled.
Epoch 397/400
538/538 - 93s - factorized_top_k/top_1_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_5_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_10_categorical_accuracy: 0.0000e+00 
- factorized_top_k/top_50_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_100_categorical_accuracy: 0.0000e+00 - loss: nan - regularization_loss: 0.0000e+00 - total_loss: nan
2022-06-14 05:23:05.912708: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 697375 of 1300000
2022-06-14 05:23:15.912724: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 1189677 of 1300000
2022-06-14 05:23:17.359999: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:228] Shuffle buffer filled.
2022-06-14 05:23:17.360079: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 1 of 1100000
2022-06-14 05:23:17.360103: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 2 of 1100000
2022-06-14 05:23:25.897982: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 1053770 of 1100000
2022-06-14 05:23:26.288890: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:228] Shuffle buffer filled.
Epoch 398/400
538/538 - 85s - factorized_top_k/top_1_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_5_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_10_categorical_accuracy: 0.0000e+00 
- factorized_top_k/top_50_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_100_categorical_accuracy: 0.0000e+00 - loss: nan - regularization_loss: 0.0000e+00 - total_loss: nan
2022-06-14 05:24:31.127985: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 590087 of 1300000
2022-06-14 05:24:41.127978: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 931083 of 1300000
2022-06-14 05:24:48.282054: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:228] Shuffle buffer filled.
2022-06-14 05:24:48.282178: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 1 of 1100000
2022-06-14 05:24:48.282209: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 2 of 1100000
2022-06-14 05:24:51.103247: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 220747 of 1100000
2022-06-14 05:24:59.776506: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:228] Shuffle buffer filled.
Epoch 399/400
538/538 - 95s - factorized_top_k/top_1_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_5_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_10_categorical_accuracy: 0.0000e+00 
- factorized_top_k/top_50_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_100_categorical_accuracy: 0.0000e+00 - loss: nan - regularization_loss: 0.0000e+00 - total_loss: nan
2022-06-14 05:26:06.147844: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 623864 of 1300000
2022-06-14 05:26:16.147848: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 926547 of 1300000
2022-06-14 05:26:23.317975: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:228] Shuffle buffer filled.
2022-06-14 05:26:23.318110: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 1 of 1100000
2022-06-14 05:26:23.318145: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 2 of 1100000
2022-06-14 05:26:26.120271: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:175] Filling up shuffle buffer (this may take a while): 284981 of 1100000
2022-06-14 05:26:35.920013: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:228] Shuffle buffer filled.
Epoch 400/400
538/538 - 92s - factorized_top_k/top_1_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_5_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_10_categorical_accuracy: 0.0000e+00 
- factorized_top_k/top_50_categorical_accuracy: 0.0000e+00 - factorized_top_k/top_100_categorical_accuracy: 0.0000e+00 - loss: nan - regularization_loss: 0.0000e+00 - total_loss: nan
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
2022-06-14 05:28:20.745514: W tensorflow/python/util/util.cc:348] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
WARNING:absl:Function `_wrapped_model` contains input name(s) phoneNo, procCd, regCnl, userEmbedding, usrBrd, usrStarLvl with unsupported characters which will be renamed to phoneno, proccd
, regcnl, userembedding, usrbrd, usrstarlvl in the SavedModel.
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:tensorflow:Model was constructed with shape (None, 50) for input KerasTensor(type_spec=TensorSpec(shape=(None, 50), dtype=tf.float32, name='pretrained_user_embedding'), name='pretra
ined_user_embedding', description="created by layer 'pretrained_user_embedding'"), but it was called on an input with incompatible shape (None,).
WARNING:absl:Found untraced functions such as query_with_exclusions while saving (showing 1 of 1). These functions will not be directly callable after loading.`
